presto-native-execution/presto_cpp/main/PeriodicMemoryChecker.h

```diff
@@ -0,0 +1,127 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <folly/experimental/FunctionScheduler.h>
+#include <cstdint>
+#include <queue>
+#include <string>
+
+namespace facebook::presto {
+/// Utility class that spawns a thread which periodically checks the memory
+/// usage and perform the following actions:
+class PeriodicMemoryChecker {
+ public:
+  struct Config {
+    /// The frequency 'PeriodicMemoryChecker' runs at.
+    uint64_t MemoryCheckerIntervalSec{1};
+
+    /// If true, starts memory limit checker to trigger memory pushback when
+    /// server is under low memory pressure.
+    bool systemMemPushbackEnabled{false};
+
+    /// Specifies the system memory limit that triggers the memory pushback if
+    /// the server memory usage is beyond this limit. This only applies if
+    /// 'systemMemPushbackEnabled_' is true.
+    uint64_t systemMemLimitBytes{0};
+
+    /// Specifies the amount of memory to shrink when the memory pushback is
+    /// triggered. This only applies if 'systemMemPushbackEnabled_' is true.
+    uint64_t systemMemShrinkBytes{0};
+
+    /// If true, starts memory limit checker that dumps the heap profile if
+    /// memory allocated via malloc exceeds 'mallocBytesUsageDumpThreshold_'.
+    /// See 'mallocCheckingCb()' for more details.
+    bool mallocMemHeapDumpEnabled{false};
+
+    /// Minimum Interval in seconds between two consecutive heap dumps.
+    size_t minHeapDumpIntervalSec{10};
+
+    /// The directory where heap profiles are dumped.
+    std::string heapDumpLogDir;
+
+    /// The prefix of heapdump file name.
+    std::string heapDumpFilePrefix;
+
+    /// Maximum number of heap dump files to maintain in rotation.
+    int maxHeapDumpFiles{10};
+
+    /// Memory (in bytes) allocated via malloc() that triggers the heap dump.
+    /// Default is 20GB.
+    size_t mallocBytesUsageDumpThreshold{20UL * 1024 * 1024 * 1024};
+  };
+
+  explicit PeriodicMemoryChecker(Config config);
+
+  void start();
+
+  void stop();
+
+ protected:
+  /// Returns current system memory usage. The returned value is used to compare
+  /// with 'Config::systemMemLimitBytes'.
+  virtual int64_t systemUsedMemoryBytes() const = 0;
+
+  /// Returns current bytes allocated by malloc. The returned value is used to
+  /// compare with 'Config::mallocBytesUsageDumpThreshold'
+  virtual int64_t mallocBytes() const = 0;
+
+  /// Callback function that is invoked by 'PeriodicMemoryChecker' at all time.
+  /// Light operations such as stats reporting can be done in this call back.
+  virtual void periodicCb() const = 0;
```

Comment:
Is is memory pushback callback? Thanks!

Reply:
No. Memory pushback call back is implemented because it is not system dependent.